code/game/objects/items/RSF.dm

```diff
@@ -8,178 +8,200 @@ RSF
 	desc = "A device used to rapidly deploy service items."
 	icon = 'icons/obj/tools.dmi'
 	icon_state = "rsf"
+	///The icon state to revert to when the tool is empty
+	var/spent_icon_state = "rsf_empty"
 	lefthand_file = 'icons/mob/inhands/equipment/tools_lefthand.dmi'
 	righthand_file = 'icons/mob/inhands/equipment/tools_righthand.dmi'
 	opacity = 0
 	density = FALSE
 	anchored = FALSE
 	item_flags = NOBLUDGEON
 	armor = list("melee" = 0, "bullet" = 0, "laser" = 0, "energy" = 0, "bomb" = 0, "bio" = 0, "rad" = 0, "fire" = 0, "acid" = 0)
+	///The current matter count
 	var/matter = 0
-	var/mode = 1
+	///The max amount of matter in the device
+	var/max_matter = 30
+	///The type of the object we are going to despense
+	var/to_despense
+	///The cost of the object we are going to despense
+	var/despense_cost = 0
 	w_class = WEIGHT_CLASS_NORMAL
+	///An associated list of atoms and charge costs. This can contain a seperate list, as long as it's associated item is an object
+	var/list/cost_by_item = list(/obj/item/reagent_containers/food/drinks/drinkingglass = 20,
+								/obj/item/paper = 10,
+								/obj/item/storage/pill_bottle/dice = 200,
+								/obj/item/pen = 50,
+								/obj/item/clothing/mask/cigarette = 10,
+								)
+	///An associated list of fuel and it's value
+	var/list/matter_by_item = list(/obj/item/rcd_ammo = 10,)
+	///A list of surfaces that we are allowed to place things on.
+	var/list/allowed_surfaces = list(/turf/open/floor, /obj/structure/table)
+	///The unit of mesure of the matter, for use in text
+	var/discriptor = "fabrication-units"
+	///The verb that describes what we're doing, for use in text
+	var/action_type = "Dispensing"
+
+/obj/item/rsf/Initialize()
+	. = ..()
+	to_despense = cost_by_item[1]
 
 /obj/item/rsf/examine(mob/user)
 	. = ..()
-	. += "<span class='notice'>It currently holds [matter]/30 fabrication-units.</span>"
+	. += "<span class='notice'>It currently holds [matter]/[max_matter] [discriptor].</span>"
 
 /obj/item/rsf/cyborg
 	matter = 30
 
 /obj/item/rsf/attackby(obj/item/W, mob/user, params)
-	if(istype(W, /obj/item/rcd_ammo))
-		if((matter + 10) > 30)
-			to_chat(user, "<span class='warning'>The RSF can't hold any more matter!</span>")
+	if(is_type_in_list(W,matter_by_item))//If the thing we got hit by is in our matter list
+		if(matter > max_matter)
```

Comment:
```suggestion
		if((matter+matter_by_item[W.type]) > max_matter)
```
I think need check that we can fully add charge in. Or we can waste full 10 charge to one point.

Reply:
yeah sure, I though the behavior was a bit odd, but I can understand why it's there.